topology/probes/vpp/vpp.go

```diff
@@ -103,268 +124,511 @@ func (p *Probe) getInterfaceVrfID(ch api.Channel, index uint32) int64 {
-	tr.AddMetadata("MAC", interfaceMAC(intf.L2Address[:intf.L2AddressLength]))
-	tr.AddMetadata("MTU", int64(intf.LinkMtu))
-	tr.AddMetadata("Speed", int64(intf.LinkSpeed))
-	state := interfaceUpDown(intf.AdminUpDown)
-	tr.AddMetadata("State", state)
-	if state != "DOWN" {
-		tr.AddMetadata("Duplex", interfaceDuplex(intf.LinkDuplex))
-	}
-	tr.AddMetadata("VrfID", p.getInterfaceVrfID(ch, intf.SwIfIndex))
-
 	return node
 }
 
 func interfaceNeedUpdate(i1, i2 *interfaces.SwInterfaceDetails) bool {
-	if (i1.LinkMtu != i2.LinkMtu) ||
-		(i1.LinkSpeed != i2.LinkSpeed) ||
-		(i1.LinkDuplex != i2.LinkDuplex) {
-		return true
-	}
-	return false
+	return i1.LinkMtu != i2.LinkMtu ||
+		i1.LinkSpeed != i2.LinkSpeed ||
+		i1.LinkDuplex != i2.LinkDuplex
 }
 
-func (p *Probe) eventAddInterface(ch api.Channel, intf *interfaces.SwInterfaceDetails) {
-	p.createOrUpdateInterface(ch, intf)
+func (p *Probe) eventAddInterface(ch api.Channel, intf *interfaces.SwInterfaceDetails, metadata graph.Metadata) *graph.Node {
+	return p.createOrUpdateInterface(ch, intf, metadata)
 }
 
 func (p *Probe) eventDelInterface(node *graph.Node) {
-	p.Ctx.Graph.Lock()
-	defer p.Ctx.Graph.Unlock()
 	if err := p.Ctx.Graph.DelNode(node); err != nil {
+		p.NotifyEvent(graph.NodeDeleted, node)
 		p.Ctx.Logger.Error(err)
 	}
 }
 
 func (p *Probe) interfaceEventsEnableDisable(ch api.Channel, enable bool) {
-	ed := uint32(0)
+	req := &interfaces.WantInterfaceEvents{PID: uint32(os.Getpid())}
 	if enable {
-		ed = uint32(1)
-	}
-	req := &interfaces.WantInterfaceEvents{
-		EnableDisable: ed,
-		PID:           uint32(os.Getpid()),
+		req.EnableDisable = 1
 	}
 	msg := &interfaces.WantInterfaceEventsReply{}
-	err := ch.SendRequest(req).ReceiveReply(msg)
-	if err != nil {
+	if err := ch.SendRequest(req).ReceiveReply(msg); err != nil {
 		p.Ctx.Logger.Error(err)
 	}
 }
 
-func (p *Probe) interfacesEvents() {
-	defer p.wg.Done()
+type memifSocket struct {
+	*memif.MemifDetails
+	filename string
+}
 
-	ch, err := p.conn.NewAPIChannel()
-	if err != nil {
-		p.Ctx.Logger.Error("API channel error: ", err)
-		return
+func (p *Probe) fetchMemif(ch api.Channel) (map[uint32]*memifSocket, error) {
+	sharedMemifSockets := make(map[uint32]*memifSocket)
+
+	req := &memif.MemifSocketFilenameDump{}
+	msg := &memif.MemifSocketFilenameDetails{}
+	reqCtx := ch.SendMultiRequest(req)
+	for {
+		if lastReply, err := reqCtx.ReceiveReply(msg); lastReply {
+			break
+		} else if err != nil {
+			return nil, err
+		}
+
+		sharedMemifSockets[msg.SocketID] = &memifSocket{filename: interfaceName(msg.SocketFilename)}
 	}
 
-	sub, err := ch.SubscribeNotification(p.notifChan, &interfaces.SwInterfaceEvent{})
+	req2 := &memif.MemifDump{}
+	msg2 := &memif.MemifDetails{}
+	reqCtx = ch.SendMultiRequest(req2)
+	for {
+		if lastReply, err := reqCtx.ReceiveReply(msg2); lastReply {
+			break
+		} else if err != nil {
+			return nil, err
+		}
+
+		sharedMemifSockets[msg2.SocketID].MemifDetails = msg2
```

Comment:
Actually there can be multiple (listener) memifs (on single VPP) that use the same socket. Master-Slave memif connection is created not only when the same socket file is used, but also when IDs (field `ID`) are the same.
In other words there can be multiple listeners with different IDs using the same socket and based on the ID of the connecting memif the match is made.